salt/config/__init__.py

```diff
@@ -3284,6 +3292,42 @@ def _cache_id(minion_id, cache_file):
         log.error('Could not cache minion ID: {0}'.format(exc))
 
 
+def eval_minion_id_func(opts):
+    '''
+    Evaluate minion ID function if 'minion_id_type' is 'func'
+    and return the result
+    '''
+    if '__minion_id_func_evaluated' in opts:
+        return opts['id']
+
+    import salt.loader as loader
+
+    # split module and function and try loading the module
+    mod_fun = opts['id']
+    mod, fun = mod_fun.split('.')
+    if not opts.get('grains'):
+        # Get grains for use by the module
+        opts['grains'] = loader.grains(opts)
+
+    try:
+        minion_id_mod = loader.raw_mod(opts, mod, fun)
+        if not minion_id_mod:
+            raise KeyError
+        # we take whatever the module returns as the minion ID
+        newid = minion_id_mod[mod_fun]()
```

Comment:
@cachedout Salt SSH has no `--local` option, sorry. This is `salt-call` does. But the whole thing is this: imagine you're bootstrapping a machine that has only SSH access. Just like we do with our SUSE Manager :wink: Now, you're supplying some data along, say your function normally creates nice IDs according to the "DNS name + Datacenter prefix", as for an example. In one DC we end up `foobar-muenchen.foo.com` and another `foobar-frankfurt.foo.com` But then where this `$PREFIX` comes from? So I am wondering if that could be considered along.